### PR TITLE
Add implementation of psp-sysctl-policy

### PR DESCRIPTION
**Pull request by viccuad**
Example requests in *test_data/* are crafted from those in the psp-allow-privilege-escalation policy.

Implementation for https://github.com/kubewarden/policy-hub/issues/31.

### CODE REVIEW

File: validate.go [hunk 1]
```diff
@@ -18,34 +37,83 @@ func validate(payload []byte) ([]byte, error) {
...
+		knownSafeSysctls.Add(knownSafeSysctlsArray[i])
+	}
+
+	// build set of prefixes from patterns of forbidden sysctls:
+	globForbiddenSysctls := mapset.NewThreadUnsafeSet()
+	for _, elem := range settings.ForbiddenSysctls.ToSlice() {
+		if strings.HasSuffix(elem.(string), "*") {
+			globForbiddenSysctls.Add(strings.TrimSuffix(elem.(string), "*"))
+		}
+	}
```
Comment (flavio) [inline validate.go:76]: I would drop/move this code somewhere else. 

I would change the internal implementation of Settings to hold:
  * `ForbiddenSysctls`
  * `ForbiddenGlobSysctls`
  * `AllowedUnsafeSysctls`

I think we could also add the `getSafeSysctls` sysctls into `AllowedUnsafeSysctls`

File: validate.go [hunk 2]
```diff
@@ -18,34 +37,83 @@ func validate(payload []byte) ([]byte, error) {
...
+
+		// if sysctl is not on the safe list nor an exception, it is forbidden:
+		if !knownSafeSysctls.Contains(sysctl) &&
+			!settings.AllowedUnsafeSysctls.Contains(sysctl) {
+			err = fmt.Errorf("sysctl %s is not on safe list, nor is in the allowedUnsafeSysctls list",
+				sysctl)
+			return false // stop iterating
+		}
+		return true // continue iterating
+	})
```
Comment (flavio) [inline validate.go:102]: I would change the block of code with this pseudo code:

```
not_allowed = data <intesection> settings.forbidden
if len(not_allowed) > 0 {
  reject(not_allowed.join(" "))
}

to_eval = data - settings.allowed
not_allowed := []
for sysctl := range to_eval {
  if match(settings.not_allowed_glob, sysctl) {
    not_allowed = append(not_allowed, sysctl)
  }
}

if len(not_allowed) > 0 {
  reject(not_allowed.join(" "))
}

accept

```

This is just a matter of personal taste. We have Set objects with all the things that are allowed and not allowed, I just happen to like to perform "set operations" as much as possible.

File: validate.go [hunk 3]
```diff
@@ -18,34 +37,83 @@ func validate(payload []byte) ([]byte, error) {
...
 
 		return kubewarden.RejectRequest(
-			kubewarden.Message(
-				fmt.Sprintf("The '%s' name is on the deny list", name)),
+			kubewarden.Message(err.Error()),
 			kubewarden.NoCode)
 	}
 
-	logger.Info("accepting ingress object")
+	logger.Info("accepting pod object")
```
Comment (flavio) [inline validate.go:117]: I would drop that. Keep in mind, the go template added logging a bit too much for my taste, but just to give an idea about how to use it. Maybe we should cleanup the template too

**Review by flavio — CHANGES_REQUESTED** [review]
Overall LGTM, just some minor changes

File: README.md [hunk 4]
```diff
@@ -1,115 +1,79 @@
...
-The unit tests can be run via a simple command:
-
-```shell
-make test
+```yaml
+forbiddenSysctls:
+- kernel.*
+- net.ipv6.conf.lo.max_addresses
+allowedUnsafeSysctls:
+- net.core.somaxconn
```
Comment (ereslibre) [inline README.md:45]: I think it would be interesting to add an `allowedUnsafeSysctls` more specific than a more generic in `forbiddenSysctls` to show that it has precedence. Like:

```yaml
forbiddenSysctls:
- net.ipv6.conf.lo.*
allowedUnsafeSysctls:
- net.ipv6.conf.lo.max_addresses
```

This instance would prevent every sysctl under `net.ipv6.conf.lo`, except `net.ipv6.conf.lo.max_addresses`, that would be allowed. There might be more interesting examples than this made-up one :)

Comment (ereslibre) [reply README.md:45]: I think we should also add a test case that accepts in this situation. A pod with `sysctl` `net.ipv6.conf.lo.max_addresses` in this context should be allowed, despite `net.ipv6.conf.lo.*` is forbidden. This would test an interesting edge case we want to preserve.

File: metadata.yml [hunk 5]
```diff
@@ -6,13 +6,53 @@ rules:
...
-    **Note well:** this can be Markdown text
+    ## Settings
+
+    The following settings are accepted:
+
+    * `forbiddenSysctls`: List of plain sysctl names or sysctl patterns (which end
+      with `*`) to be forbidden. You can forbid a combination of safe and unsafe
+      sysctls in the list. To forbid setting any sysctls, use * on its own.
+    * `allowedUnsafeSysctls`: List of plain sysctl names that can be used in Pods.
+      `*` cannot be used.
```
Comment (ereslibre) [inline metadata.yml:29]: I think this documentation has to specify the priority. `allowedUnsafeSysctls` takes priority over `forbiddenSysctls` if it's more specific.

**Review by ereslibre — COMMENTED** [review]
Some comments, thanks @viccuad!

File: README.md [hunk 6]
```diff
@@ -1,115 +1,79 @@
...
+Replacement for the Kubernetes Pod Security Policy that controls the usage of
+sysctls.
 
-## Introduction
+Linux Kernel sysctls are grouped into safe and unsafe sets. A safe sysctl must
+be properly isolated between pods on the same node, and are properly namespaced
+by the kernel. 
 
-This repository contains a working policy written in Go.
+All safe sysctls are enabled by default in k8s.
```
[the remark this answers is not in the document]
Comment (viccuad) [reply README.md]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/b97f92b9f991fc10fa28cca6842953a66a1dfecc.

File: README.md [hunk 7]
```diff
@@ -1,115 +1,79 @@
...
+All unsafe sysctls are disabled by default and must be explicitly allowed on a
+per-node or per-pod basis.
 
-The policy looks at the `name` of a Kubernetes resource and rejects the request
-if the name is on a deny list.
+As the deprecated analogous Kubernetes PSP, this policy validates which sysctls
+can get set in pods by specifying lists of sysctls or sysctl patterns to be
+allowed or forbidden. One can then modify the securityContext of Pods to make
```
[the remark this answers is not in the document]
Comment (viccuad) [reply README.md]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/b97f92b9f991fc10fa28cca6842953a66a1dfecc.

File: README.md [hunk 8]
```diff
@@ -1,115 +1,79 @@
...
-the actual request received by Kubernetes.
-However it's still possible to write a Kubewarden policy by using some 3rd party
-libraries.
+The following settings are accepted:
 
-This is a list of libraries that can be useful when writing a Kubewarden
-policy:
+* `forbiddenSysctls`: List of plain sysctl names or sysctl patterns (which end
+  with `*`) to be forbidden. You can forbid a combination of safe and unsafe
+  sysctls in the list. To forbid setting any sysctls, use * on its own.
```
[the remark this answers is not in the document]
Comment (viccuad) [reply README.md]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/b97f92b9f991fc10fa28cca6842953a66a1dfecc.

File: metadata.yml [hunk 9]
```diff
@@ -6,13 +6,53 @@ rules:
...
+    kind: Pod
+    metadata:
+      name: sysctl-example
+    spec:
+      securityContext:
+        sysctls:
+        - name: net.core.somaxconn
+          value: "1024"
+    ...
+    ```
```
[the remark this answers is not in the document]
Comment (viccuad) [reply metadata.yml:58]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/7ecf7bb2233c5b8cae76762c426d2fa28448942c.

File: README.md [hunk 10]
```diff
@@ -1,115 +1,79 @@
...
+kind: Pod
+metadata:
+  name: sysctl-example
+spec:
+  securityContext:
+    sysctls:
+    - name: kernel.shm_rmid_forced
+      value: "0"
+    - name: kernel.msgmax
+      value: "65536"
```
[the remark this answers is not in the document]
Comment (viccuad) [reply README.md:77]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/7ecf7bb2233c5b8cae76762c426d2fa28448942c.

File: validate.go [hunk 11]
```diff
@@ -18,34 +37,83 @@ func validate(payload []byte) ([]byte, error) {
 
 	logger.Info("validating request")
 
+	if kind := gjson.GetBytes(payload, "request.object.kind"); kind.String() != "Pod" {
+		return kubewarden.RejectRequest(
+			kubewarden.Message("object is not of kind Pod: rejecting request"),
+			kubewarden.Code(421))
+	}
+
```
[the remark this answers is not in the document]
Comment (viccuad) [reply validate.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/14fd3596a8ec7f57a8b6c0102ecb44fc7a55a8f0.

File: e2e.bats [hunk 12]
```diff
@@ -1,19 +1,32 @@
...
 
   # this prints the output when one the checks below fails
   echo "output = ${output}"
 
   # request rejected
   [ "$status" -eq 0 ]
   [ $(expr "$output" : '.*allowed.*false') -ne 0 ]
-  [ $(expr "$output" : ".*The 'tls-example-ingress' name is on the deny list.*") -ne 0 ]
+  [ $(expr "$output" : ".*object is not of kind Pod: rejecting request.*") -ne 0 ]
 }
```
[the remark this answers is not in the document]
Comment (viccuad) [reply e2e.bats]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/14fd3596a8ec7f57a8b6c0102ecb44fc7a55a8f0.

File: validate.go [hunk 13]
```diff
@@ -18,34 +37,83 @@ func validate(payload []byte) ([]byte, error) {
...
 	data := gjson.GetBytes(
 		payload,
-		"request.object.metadata.name")
+		"request.object.spec.securityContext.sysctls")
 
 	if !data.Exists() {
-		logger.Warn("cannot read object name from metadata: accepting request")
+		logger.Warn("pod doesn't specify sysctls: accepting request")
```
[the remark this answers is not in the document]
Comment (viccuad) [reply validate.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/74ce514b0e87f365c2473773e7d04b22044d254e.

File: validate.go (hunk 3 above)
Comment (viccuad) [reply validate.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/74ce514b0e87f365c2473773e7d04b22044d254e.
Opened https://github.com/kubewarden/go-policy-template/pull/8.

File: settings.go [hunk 14]
```diff
@@ -6,73 +6,115 @@ import (
...
 	if err != nil {
-		return []byte{}, err
+		return kubewarden.RejectSettings(
+			kubewarden.Message(fmt.Sprintf("provided settings are not valid: %v", err)))
 	}
 
-	if settings.Valid() {
+	valid, err := settings.Valid()
+	if valid {
 		logger.Info("accepting settings")
```
[the remark this answers is not in the document]
Comment (viccuad) [reply settings.go]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/74ce514b0e87f365c2473773e7d04b22044d254e.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (viccuad) [reply README.md]: Resolving conversations because there's a bunch (normally I would leave it to reviewers to mark as resolved).

File: metadata.yml (hunk 5 above)
Comment (viccuad) [reply metadata.yml:29]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/ae703568bd79eada02a87e9f7350222ccbc6a446.

File: README.md (hunk 4 above)
Comment (viccuad) [reply README.md]: Changed unit and e2e tests to check for precedence in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/ca24522f6bae4d377951496788d23b4475bd285e.
Changed readme in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/ae703568bd79eada02a87e9f7350222ccbc6a446.

File: metadata.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (viccuad) [reply metadata.yml:81]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/520e46bcc7efcd4583086d8828ba4493159d02a2.

File: metadata.yml (hunk 5 above)
Comment (viccuad) [reply metadata.yml]: Done in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/520e46bcc7efcd4583086d8828ba4493159d02a2.

File: validate.go (hunk 1 above)
Comment (viccuad) [reply validate.go]: A priori sounds like a good idea at least for clarity. But then the Settings struct will not match the json (nice if we go with encoding/json). We could built an internal struct with the `forbidden`, `forbiddenGlob`, `safe` sets, instead of creating them at the beginning of evaluation. But I prefer to leave the current `knownSafeSysctls` because then We can provide a better message when rejecting.

Simplified `knownSafeSysctls` creation by using a set instead of array in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/8b03dfdf2592e8e5c2bc6b6c32ba929bde7bdb61 (I was using an array to signify that they are "constant", but it's not really needed).

I don't think it leads to a faster implementation though; right now we are iterating through the pod sysctls and if we find one sysctl that is not good, we stop and reject. With the sets, we would first perform set operations (which, given that the are implemented with maps is going to need the same iterations), and then iterate nevertheless.

File: validate.go (hunk 2 above)
Comment (viccuad) [reply validate.go]: This does sound clearer, but I don't think it leads to a faster implementation though; right now we are iterating through the pod sysctls and if we find one sysctl that is not good, we stop and reject. With the sets, we would first perform set operations (which, given that the are implemented with maps is going to need the same iterations), and then iterate nevertheless, as we need to check against each blob prefix.

**Comment by viccuad** [comment]
Thanks both for the reviews! :). The CI is still not running, maybe because I set the repo to "Require approval for all outside collaborators", which sounds sane given that we have secrets for ghcr.io. I think those still need to be added, but I don't know them myself.

File: README.md [hunk 15]
```diff
@@ -1,115 +1,82 @@
...
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: sysctl-example
+spec:
+  securityContext:
+    sysctls:
+    - net.ipv6.conf.lo.max_addresses
```
Comment (ereslibre) [inline README.md:58]:
```suggestion
    - name: net.ipv6.conf.lo.max_addresses
```

File: README.md [hunk 16]
```diff
@@ -1,115 +1,82 @@
...
+apiVersion: v1
+kind: Pod
+metadata:
+  name: sysctl-example
+spec:
+  securityContext:
+    sysctls:
+    - name: kernel.shm_rmid_forced
+      value: "0"
+    - net.ipv6.conf.lo.mtu
```
Comment (ereslibre) [inline README.md:79]:
```suggestion
    - name: net.ipv6.conf.lo.mtu
```

File: metadata.yml [hunk 17]
```diff
@@ -6,13 +6,76 @@ rules:
...
+
+    ``` yaml
+    apiVersion: v1
+    kind: Pod
+    metadata:
+      name: sysctl-example
+    spec:
+      securityContext:
+        sysctls:
+        - net.ipv6.conf.lo.max_addresses
```
Comment (ereslibre) [inline metadata.yml:57]:
```suggestion
        - name: net.ipv6.conf.lo.max_addresses
```

File: metadata.yml [hunk 18]
```diff
@@ -6,13 +6,76 @@ rules:
...
+    apiVersion: v1
+    kind: Pod
+    metadata:
+      name: sysctl-example
+    spec:
+      securityContext:
+        sysctls:
+        - name: kernel.shm_rmid_forced
+          value: "0"
+        - net.ipv6.conf.lo.mtu
```
Comment (ereslibre) [inline metadata.yml:78]:
```suggestion
        - name: net.ipv6.conf.lo.mtu
```

**Review by ereslibre — APPROVED** [review]
LGTM, just pointed out some minor typos :)

**Comment by ereslibre** [comment]
> Thanks both for the reviews! :). The CI is still not running, maybe because I set the repo to "Require approval for all outside collaborators", which sounds sane given that we have secrets for ghcr.io. I think those still need to be added, but I don't know them myself.

I think is fine to run CI here. We don't have secrets per-se, we are using `GITHUB_TOKEN` that is automatically filled by GitHub, and we use that to [sign-in in the GitHub registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry).

File: README.md (hunk 15 above)
Comment (viccuad) [reply README.md]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

File: README.md (hunk 16 above)
Comment (viccuad) [reply README.md]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

File: metadata.yml (hunk 17 above)
Comment (viccuad) [reply metadata.yml]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

File: metadata.yml (hunk 18 above)
Comment (viccuad) [reply metadata.yml]: doh.. fixed in https://github.com/kubewarden/psp-sysctl-policy/pull/1/commits/729fcd8e9e3deeb5c0b8010ff27f55d34c4acc91.

**Review by flavio — APPROVED** [review]
LGTM, thanks for the changes!
